libraries/SITL/SIM_GPS.cpp

```diff
@@ -1017,6 +1017,341 @@ uint32_t GPS::CalculateBlockCRC32(uint32_t length, uint8_t *buffer, uint32_t crc
     return( crc );
 }
 
+uint8_t GPS::serialize_bitmask(const Bitmask<8>& mask)
+{
+    uint8_t ret = 0;
+    for (uint8_t i = 0 ; i < 8 ; i++)
+    {
+        if (mask.get(i)) {
+            ret |= 1 << i; 
+        }
+    }
+
+    return ret;
+}
+
+void GPS::update_gsof(const struct gps_data *d)
+{
+    // https://receiverhelp.trimble.com/oem-gnss/index.html#GSOFmessages_TIME.html?TocPath=Output%2520Messages%257CGSOF%2520Messages%257C_____25
+    constexpr uint8_t GSOF_POS_TIME_TYPE { 0x01 };
+    constexpr uint8_t GSOF_POS_TIME_LEN { 0x0A };
+     // TODO magic number until SITL supports GPS bootcount based on GPSN_ENABLE
+    const uint8_t bootcount = 17;
+
+    // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Flags.html#Position%20flags%201
+    enum class pos_flags_1_fields : uint8_t {
+        NEW_POSITION = 0,
+        CLOCK_FIX_CALULATED = 1,
+        HORIZ_FROM_THIS_POS = 2,
+        HEIGHT_FROM_THIS_POS = 3,
+        RESERVED_4 = 4,
+        LEAST_SQ_POSITION = 5,
+        RESERVED_6 = 6,
+        POSITION_L1_PSEUDORANGES = 7
+    };
+    Bitmask<8> pos_flags_1;
```

Comment:
Just use a `uint8_t`

Reply:
Isn't this the point of the bitmask class?

Reply:
I'm concerned about the readability of it.

Reply:
I'll give it a whirl.

Reply:
I'd like to understand how you want me to refactor the bitfields:

Other examples in `master` are some magic numbers and comments. 
1. `pvt.flags = 0b10000011; // carrsoln=fixed, psm = na, diffsoln and fixok`

2. `svinfo.globalFlags = 4; // u-blox 8/M8`

3. `relposned.flags = gnssFixOK | diffSoln | carrSolnFixed | isMoving | relPosValid | relPosHeadingValid;`

4. `svinfo.sv[i].flags = (i < _sitl->gps_numsats[instance]) ? 0x7 : 0x6; // sv used, diff correction data, orbit information`


The only one with conditional logic is the last one, and I don't understand what `0x7` vs `0x6` is. If there is a way I can preserve the notion that bit 1 of velocity flags means `CONSECUTIVE_MEASUREMENTS`, that would be ideal. Option 3 seems most likely what I need in the `ubx` code.

If it helps, I could change those into functions so they are a bit more separated from the rest of the code, however, that isn't a pattern in the current simulation code - it's all baked into one function currently.

Reply:
Something like this:
```
diff --git a/libraries/SITL/SIM_GPS.cpp b/libraries/SITL/SIM_GPS.cpp
index 7a27606fdca..6cfa89f14db 100644
--- a/libraries/SITL/SIM_GPS.cpp
+++ b/libraries/SITL/SIM_GPS.cpp
@@ -1039,25 +1039,26 @@ void GPS::update_gsof(const struct gps_data *d)
     const uint8_t bootcount = 17;
 
     // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Flags.html#Position%20flags%201
-    enum class pos_flags_1_fields : uint8_t {
-        NEW_POSITION = 0,
-        CLOCK_FIX_CALULATED = 1,
-        HORIZ_FROM_THIS_POS = 2,
-        HEIGHT_FROM_THIS_POS = 3,
-        RESERVED_4 = 4,
-        LEAST_SQ_POSITION = 5,
-        RESERVED_6 = 6,
-        POSITION_L1_PSEUDORANGES = 7
+    enum pos_flags_1_fields {
+        NEW_POSITION = 1U << 0,
+        CLOCK_FIX_CALULATED = 1U << 1,
+        HORIZ_FROM_THIS_POS = 1U << 2,
+        HEIGHT_FROM_THIS_POS = 1U << 3,
+        RESERVED_4 = 1U << 4,
+        LEAST_SQ_POSITION = 1U << 5,
+        RESERVED_6 = 1U << 6,
+        POSITION_L1_PSEUDORANGES = 1U << 7
     };
-    Bitmask<8> pos_flags_1;
-    pos_flags_1.set(static_cast<uint8_t>(pos_flags_1_fields::NEW_POSITION));
-    pos_flags_1.set(static_cast<uint8_t>(pos_flags_1_fields::CLOCK_FIX_CALULATED));
-    pos_flags_1.set(static_cast<uint8_t>(pos_flags_1_fields::HORIZ_FROM_THIS_POS));
-    pos_flags_1.set(static_cast<uint8_t>(pos_flags_1_fields::HEIGHT_FROM_THIS_POS));
-    pos_flags_1.set(static_cast<uint8_t>(pos_flags_1_fields::RESERVED_4));
-    pos_flags_1.set(static_cast<uint8_t>(pos_flags_1_fields::LEAST_SQ_POSITION));
-    pos_flags_1.clear(static_cast<uint8_t>(pos_flags_1_fields::RESERVED_6));
-    pos_flags_1.set(static_cast<uint8_t>(pos_flags_1_fields::POSITION_L1_PSEUDORANGES));
+    const uint8_t pos_flags_1 =
+        uint8_t(pos_flags_1_fields::NEW_POSITION) |
+        uint8_t(pos_flags_1_fields::CLOCK_FIX_CALULATED) |
+        uint8_t(pos_flags_1_fields::HORIZ_FROM_THIS_POS) |
+        uint8_t(pos_flags_1_fields::HEIGHT_FROM_THIS_POS) |
+        uint8_t(pos_flags_1_fields::RESERVED_4) |
+        uint8_t(pos_flags_1_fields::LEAST_SQ_POSITION) |
+        // uint8_t(pos_flags_1_fields::RESERVED_6)) |
+        uint8_t(pos_flags_1_fields::POSITION_L1_PSEUDORANGES)
+        ;
 
     // https://receiverhelp.trimble.com/oem-gnss/GSOFmessages_Flags.html#Position%20flags%202
     enum class pos_flags_2_fields : uint8_t {
@@ -1106,7 +1107,7 @@ void GPS::update_gsof(const struct gps_data *d)
         uint8_t initialized_num = bootcount; 
     } pos_time;
     static_assert(sizeof(gsof_pos_time) - (sizeof(gsof_pos_time::OUTPUT_RECORD_TYPE) + sizeof(gsof_pos_time::RECORD_LEN)) == GSOF_POS_TIME_LEN);
-    pos_time.pos_flags_1 = serialize_bitmask(pos_flags_1);
+    pos_time.pos_flags_1 = pos_flags_1;
     pos_time.pos_flags_2 = serialize_bitmask(pos_flags_2);
     pos_time.time_week_ms = htobe32(gps_tow.ms);
     pos_time.time_week = htobe16(gps_tow.week);
```

This sort of thing can also be done with union of struct and byte, but only clouds things here and is chiefly useful when pulling things apart rather than assembling them.

Reply:
This causes conflicting enum definitions with the reserved fields because they are named the same. I guess I could take them out, but then they can't be set.

Reply:
> This causes conflicting enum definitions with the reserved fields because they are named the same. I guess I could take them out, but then they can't be set, and Trimble's docs say that some of them should be set, which I did in the sim.

I have no idea what any of that means.  The code compiled with this change in.  I did remove the `class` from `enum class` when I shouldn't have, however.  The code compiles with the patch as-is, and with it reverted to `enum class` and `: uint8_t`.

Reply:
Even with the recommended changes, it's only a few lines shorter. I could take another approach, remove the `clear` calls to options that are already 0, and add a utility to cast the enum classes to `uint8_t`'s

Reply:
There's two `RESERVED_4`'s in it. 

Have a look at the error on my other branch:     `ryanf55:gsof-sim-no-bitmask`

Reply:
Fixed, went back to `enum class` with the static_casts and it works. Less code than before, just as readable.